### PR TITLE
Hide password in logs

### LINKED ISSUE #233: Possible Security Issue: AUTH credentials visible in log
I was fooling around with a branch, and decided to activate logging.

Found out that all `AUTH` commands are logged as-is, without sanitizing. Even if logging level set to `INFO`.

Not a problem if using multistep authentication (i.e., `AUTH mech` then wait for `334 ` then respond; the `_auth_interact` method does not log client response), but a problem if using one-step authentication (e.g., `AUTH PLAIN <encoded_username_and_password>`)

I fear this might cause to inadvertent leak of sensitive information. For example, if log is being forwarded to an external log collector (e.g., CloudWatch Log on AWS, DataDog, etc.)

I will submit an URGENT PR to fix this.

### PR DESCRIPTION
**Pull request by pepoluan**
## What do these changes do?

Sanitize command line before logging.

This prevents leakage of sensitive information.

## Are there changes in behavior for the user?

Only if user parses the log output.

If user _absolutely_ needs to capture the SMTP credentials, user can:

  1. Override the `auth_*` methods and/or set `log_client_response=True` on calls to `SMTP._auth_interact()` (for custom, non-built-in AUTH mechanisms), or
  2. Perform the logging within the `auth_callback` function

## Related issue number

Closes #233 

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] tox testenvs have been executed in the following environments:
   - [x] Windows 10 (via PyCharm tox runner): `(ALL)`
   - [x] Windows 10 (via PSCore 7.1.0): `(ALL)`
   - [x] Windows 10 (via Cygwin): `qa,py36-{nocov,cov}`
   - [x] Ubuntu 18.04 on WSL 1.0: `(ALL)` + `pypy3-{nocov,cov,diffcov}`
   - [x] Ubuntu 18.04 on VBox: `(ALL)` + `pypy3-{nocov,cov,diffcov}`
   - [x] Ubuntu 20.04 on VBox: `(ALL)` + `pypy3-{nocov,cov,diffcov}`
   - [x] FreeBSD 12.2 [1] on VBox: `(ALL)` + `pypy3-{nocov,cov,diffcov}`
- [x] Documentation reflects the changes
- [x] Add a news fragment into the `NEWS.rst` file

[1] [12.1 is being EOL-ed](https://www.freebsd.org/security/index.html#sup). Side benefit: Since I decided to just wipe my VM and reinstall from scratch, lo and behold! Now all `pypy3` tests work! And the icing on the cake is that the whole tox suite runs much faster than before!

### CODE REVIEW

File: aiosmtpd/smtp.py
Comment (pepoluan) [inline aiosmtpd/smtp.py:34]: For better representation in the logs.

File: aiosmtpd/smtp.py
Comment (pepoluan) [inline aiosmtpd/smtp.py:752]: Come to think of it, logging AUTH interaction _might_ be useful in some situations.

So rather than hardcoding no logging, we'll provide a default-disabled logging here.

File: aiosmtpd/testing/helpers.py
Comment (pepoluan) [inline aiosmtpd/testing/helpers.py:70]: Previous logic stops on mismatch of code. Hiding possibly important info contained in the message.

So I rewrite the comparo to directly compare the tuples; no more hidden info :-)

**Review by pepoluan — COMMENTED** [review]
Some comments have been left for possibly interesting points.

**Comment by pepoluan** [comment]
I am so happy that I took the 'nuclear' option and just nuked my FreeBSD 12.1 VM instead of upgrading it to 12.2.

Now all testenvs work, including `pypy3-cov` and `pypy3-diffcov`. And the tests run _much_ faster!

I'm not blaming 12.1 for the problems I've experienced; I haven't used FreeBSD for quite some time before I decided to add one as a test system, so I might have inadvertently b0rked my 12.1 VM. When I recreated the VM with 12.2 from scratch, my FreeBSD knowledge had been refreshed quite a bit so I likely set the new VM the right way.

File: aiosmtpd/smtp.py
Comment (pepoluan) [inline aiosmtpd/smtp.py:183]: We can extend the `b"*"` string to 20 characters if we need more security 😝

**Comment by pepoluan** [comment]
Since this is kinda urgent, after 2x24 I'll squamerge if no review.

File: aiosmtpd/smtp.py
Comment (waynew) [inline aiosmtpd/smtp.py:464]: **suggestion**: Would it make more sense to create a `sanitized_log_debug`/`info`, rather than depend on remembering to call `sanitize`?

**Review by waynew — APPROVED** [review]
I'm :+1: on this, not sure if we could improve things a little by creating some dedicated logging functions for logging sensitive info? :shrug:

File: aiosmtpd/smtp.py
Comment (pepoluan) [reply aiosmtpd/smtp.py:464]: Well, that would only change the need of remembering to call `sanitize` with the need to call `sanitized_log_*`, so I don't really see the benefits.

Comment (waynew) [reply aiosmtpd/smtp.py:464]: It's definitely a small thing, but it boils down to this: 

```
log.debug('blahahahahahblaksj %s', whatever)
log.debug('bargbargbargbarg arg %s', sanitize (bob))
```

Is easy to miss that one has the sanitize call and the other is missing. Especially in a huge block of code.

It's even worse with a multitude of parameters.

Is it the end of the world? No. But are there tangible benefits that make writing and reviewing easier? Sure 😁

Basically any time you see a logging call that sends sensitive data, it should be calling the sensitive log function.

So... A small thing that I don't think should block the PR. But it would be a useful thing 🙂

Comment (pepoluan) [reply aiosmtpd/smtp.py:464]: Hmmm, you do have a point there.

Let me mull it over... I definitely don't want to make a too-complex sanitizing system... but I think I have some inkling on how to implement it nicely...

Comment (waynew) [reply aiosmtpd/smtp.py:464]: I think I was actually reviewing this first on mobile :upside_down_face: 

I'm OK if we want to merge this first and then do that in another PR. Otherwise, the thought I had during discussion is that we could either:

a) create a logger/handler implementation, so it's literally created like:

```
sanitized_logger = CustomLogger()
...
sanitized_logger.debug(...)
```

b) create a function/group of functions that live somewhere, so

```
from utils import sanitized_debug
...
sanitized_debug(...)
```

or similar.

Comment (pepoluan) [reply aiosmtpd/smtp.py:464]: I much prefer option (b), it's less complex and can be implemented cleanly.

Rather than wrapping all calls, I think of going like this:

```python
def sanitized_log(func, msg, *args):
    msg = sanitize(msg)
    gerard(log, func)(msg, *args)
```

Where sanitizing is needed, we change the call

```
log.debug(...)
```

to

```
sanitized_log("debug", ...)
```

What do you think?

Comment (waynew) [reply aiosmtpd/smtp.py:464]: Maybe just `sanitized_log(log.debug, ...)` then skip the getattr call? Otherwise I'm +1 for that approach

Comment (pepoluan) [reply aiosmtpd/smtp.py:464]: That's also good!

I'll implement that asap.

File: aiosmtpd/tests/test_smtp.py
Comment (pepoluan) [inline aiosmtpd/tests/test_smtp.py:65]: If env var `AIOSMTPD_TESTLOGFILE` is set, then `SMTP` logging for this test module will also be written to that file. It's optional, but sometimes... entertainig :-)

File: aiosmtpd/tests/test_smtp.py
Comment (pepoluan) [inline aiosmtpd/tests/test_smtp.py:111]: The only reason is to test / cover the `log_client_response=True` path of `SMTP.challenge_auth()`

I was a bit lazy to write yet-another-auth-mechanism (because I'll then have to edit several places), so I just 'repurpose' existing one.

File: aiosmtpd/tests/test_smtp.py
Comment (pepoluan) [inline aiosmtpd/tests/test_smtp.py:272]: This is so that in the TESTLOGFILE, the log lines will be tidy. `MockedPeer` instead of `<Mock object bla bla bla id="12345567" etc yadda yadda>`

File: aiosmtpd/tests/test_smtp.py
Comment (pepoluan) [inline aiosmtpd/tests/test_smtp.py:991]: Ensure that log is sanitized.

So in the future, if a maintainer somehow forgets that those particular lines in `smtp.SMTP` needs sanitizing, the test case will fail.

File: aiosmtpd/smtp.py
Comment (pepoluan) [inline aiosmtpd/smtp.py:794]: No sanitizing here but we do raise a pair of warnings. For _every single time_ `challenge_auth` is invoked with `log_client_response=True`.

**Comment by pepoluan** [comment]
I think we're all set. I'll wait an additional 24 hours before squamerging.

File: aiosmtpd/tests/test_smtp.py
Comment (waynew) [reply aiosmtpd/tests/test_smtp.py:991]: **praise**: I *love* providing tests for future maintainers. Usually because I'm that future maintainer :trollface:

**Review by waynew — APPROVED** [review]
:rocket:

**Comment by pepoluan** [comment]
@waynew @warsaw I have published a new "GitHub Release". Need your help to push this to PyPI, because I'm not yet a maintainer for aiosmtpd there.

**Pull request merged**
